Add an option to template delay_on/off in template binary sensor

mLupine opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR introduces the possibility to use templates in `delay_on` and `delay_off` parameters of a Template Binary Sensor. 

The returned value of such template can be a string, in a form of `hh:mm:ss`, or a dictionary, using the newly introduced rich template return types.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->


```yaml
input_number:
  delay:
    initial: 3
    min: 0
    max: 59
    step: 1

binary_sensor:
  - platform: template
    sensors:
      test_template:
        friendly_name: Delay template
        delay_off: >-
          {{ ({ "seconds": states('input_number.delay') }) }}
        value_template: >-
          {{ is_state('input_boolean.test', 'on') }}
```


```yaml
binary_sensor:
  - platform: template
    sensors:
      test_template:
        friendly_name: Delay template
        delay_off: >-
          {% if is_state('switch.night_mode', 'off') %}
              00:00:15
          {% else %}
              00:00:35
          {% endif %}
        value_template: >-
          {{ is_state('input_boolean.test', 'on') }}

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15654

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mLupine committed:
Add an option to template delay_on/off in template binary sensor

mLupine committed:
Fix formatting

bdraco reviewed (changes requested):
The test is only testing a value that won't change which means the test isn't covering the new functionality. Please add tests that follow the example provided in the description with using `input_number.delay`

mountainsandcode reviewed (changes requested):
If we want to go ahead with this, I would suggest that we add template support to delay_on and delay_off directly, rather than create separate fields. I see two advantages with that approach

1. If we have two separate attributes, this means its unclear which takes precedence over the other (if both are present)
2. It makes things less clear for the average user, since many other fields support templates directly

In fact, from the title of the PR, I would have assumed that this is what the PR is trying to achieve

mLupine committed:
Add tests with delay based on input number

mLupine committed:
Add missing statements

mLupine commented:
Hi @bdraco, I've added the missing tests.

@mountainsandcode, I don't agree that templates should be supported in the original fields. While there are indeed some parameters that support it, they are mostly the ones that control the resource's value. There are also multiple parameters that use separate fields, suffixed with `_template` and a quick search through the code shows that it might be, in fact, the more common approach, thus making my solution more consistent with other resources.

bdraco commented:
We have actually been moving away from adding _template to existing fields and collapsing them back so we don't have multiple config options now that the code has been optimized to handle static strings

mLupine commented:
@bdraco Is there any example on how it's implemented in other resources? I'm actually struggling to find an example of how it looks like in the code anywhere in the `template` component.

bdraco commented:
Check the recent release notes about automations

mLupine commented:
@bdraco All I can find is unifying `data` with `data_template` and `service` with `service_template`. If I'm not wrong, there's no _component-specific_ attribute that accepts templates without the suffix and I'm not sure if that change is the right place to start doing that.

bdraco commented:
This might be helpful 

https://community.home-assistant.io/t/stop-requiring-magic-to-enable-templates/219523?u=bdraco

mLupine committed:
Accept templates in the default parameters

mLupine committed:
Fix import order

mLupine committed:
Remove unused import

mLupine commented:
@bdraco Changes implemented, templates are accepted in `delay_on` and `delay_off` now. I removed the newly added `*_template` fields as they're not needed for backwards compatibility since it was never merged, and also updated the docs PR.

bdraco reviewed (approved):
LGTM

bdraco commented on homeassistant/components/template/binary_sensor.py:
```diff
@@ -133,15 +133,33 @@ def __init__(
 
         self.add_template_attribute("_state", self._template, None, self._update_state)
 
+        try:
```
Actually we should probably check to see if `self._delay_on_raw is not None` before doing the try block.

mLupine replied:
I moved the if condition from the `except` block to wrap around the entire `try`. The overall behaviour of the code remains the same but it saves a few function calls 😉

mLupine committed:
Reformat conditions

bdraco reviewed (approved):
Nice

bdraco commented:
Nice first contribution 🐬 👍 🥇
